Issue 3051 - bpx activation energies

darryl-ad opened the pull request:
# Description

Fixes a bug where `E_a` was used as the alias for multiple different activation energies in `bpx`, causing them to all take the final value. Also fixes bug where `reaction rate constant activation energy` was always set as 0.
- Assigned unique variable names
- Edited naming convention for better clarity
- Updated variable name `reaction rate activation energy` -> `reaction rate constant activation energy`  in `bpx` parameter search.

Fixes # [3051](https://github.com/pybamm-team/PyBaMM/issues/3051) and # [3225](https://github.com/pybamm-team/PyBaMM/issues/3225)

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

darryl-ad commented:
Added an additional commit to fix https://github.com/pybamm-team/PyBaMM/issues/3225 here. I was concerned separate PRs would cause a merge conflict.

tinosulzer reviewed (changes requested):
Looks good, can you add tests for coverage? 

You could also reduce code duplication by defining the arrhenius function up front and just calling it everywhere e.g.
```python
def arrhenius(Ea, T):
     return exp(Ea / constants.R * (1 / T_ref - 1 / T))
```
and later 
```python
return arrhenius(Ea_D_e, T) * ...
```

darryl-ad commented:
- refactored code as suggested
- added a simple unit test for coverage, and to validate arrhenius calculations

I've not made a test before, so please double check it's ok!

See separate discussion here https://github.com/pybamm-team/PyBaMM/issues/3225 about full testing requirements for BPX.

brosaplanella commented:
The test looks good to me. The issues with the coverage seem to be that for various parameters it never reaches the `else` statement. The two previous `if` statements check for callables and data, so I guess the `else` should be for constant values, but I am not sure why `test_constant_functions` is not covering those lines.

darryl-ad commented:
I agree, I'm not sure why `test_constant_functions` no longer covers those lines

brosaplanella commented:
Can you run it in debug mode and check why the constant case does not reach the else?

darryl-ad commented:
Sorry not sure what you mean by debug mode. However I tested it with print statements and the constant case does reach the else

brosaplanella commented:
I think the issue is that in `test_constant_functions` the parameter values are created from BPX but then not used. Therefore, the header of the function is covered (as the tests goes inside the if statement) but the return line doesn't as it is not get called. I think that if you add some additional tests after line 163 which check that the values are as expected it should be fine.

darryl-ad commented:
@brosaplanella I've added a test that I think addresses your previous comment, thanks!

brosaplanella commented on CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@
 - Error generated when invalid parameter values are passed. ([#3132](https://github.com/pybamm-team/PyBaMM/pull/3132))
 - Thevenin() model is now constructed with standard variables: `Time [s], Time [min], Time [h]` ([#3143](https://github.com/pybamm-team/PyBaMM/pull/3143))
 - Fix SEI Example Notebook ([#3166](https://github.com/pybamm-team/PyBaMM/pull/3166))
+- Fixed bug causing incorrect activation energies using `create_from_bpx()` ([#3242](https://github.com/pybamm-team/PyBaMM/pull/3242))
```
Can you move this up so they PRs are in decreasing order?

darryl-ad replied:
done!

brosaplanella reviewed (approved):
Looks good to me! I couldn't tell why the doc tests are failing though, @agriyakhetarpal do you have any insight?

agriyakhetarpal commented:
There is a reST syntax warning/error in the installation instructions for WSL. I am adding a link to the erroneous lines because I cannot suggest changes directly here:

https://github.com/pybamm-team/PyBaMM/blob/develop/docs/source/user_guide/installation/windows-wsl.rst?plain=1#L24-L36

Removing the extra colons in points 3 and 4 will fix it

brosaplanella commented:
Thanks Agriya! @darryl-ad, can you resolve the conflict with CHANGELOG so tests run, and do the small fix Agriya suggested?

darryl-ad commented:
@brosaplanella I've resolved the CHANGELOG conflict. All tests are passing on my end now. The erroneous lines in the docs that Agriya suggested aren't present in this branch - do I need to create an issue and branch for it?

brosaplanella commented on CHANGELOG.md:
```diff
@@ -8,8 +8,8 @@
 
 ## Bug fixes
-
 - Fixed bug causing incorrect activation energies using `create_from_bpx()` ([#3242](https://github.com/pybamm-team/PyBaMM/pull/3242))
```
Can you move this line down so the PRs are sorted in decreasing order? Happy to merge after that

darryl-ad replied:
Done!
